moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,365 @@
+      val parameterTypes = method.genericParameterTypes
+      val parameterAnnotations = method.parameterAnnotations
+      val methodSignatureIncludesJsonWriterAndJsonAdapter = parameterTypes.size >= 2 &&
+        parameterTypes[0] == JsonWriter::class.java &&
+        returnType == Void.TYPE &&
+        parametersAreJsonAdapters(2, parameterTypes)
+      return when {
+        methodSignatureIncludesJsonWriterAndJsonAdapter -> {
+          // void pointToJson(JsonWriter jsonWriter, Point point) {
+          // void pointToJson(JsonWriter jsonWriter, Point point, JsonAdapter<?> adapter, ...) {
+          val qualifierAnnotations = parameterAnnotations[1].jsonAnnotations
+          object : AdapterMethod(
+            adaptersOffset = 2,
+            type = parameterTypes[1],
+            parameterCount = parameterTypes.size,
+            annotations = qualifierAnnotations,
+            adapter = adapter,
+            method = method,
+            nullable = true
+          ) {
+            override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+              invoke(writer, value)
+            }
+          }
+        }
+        parameterTypes.size == 1 && returnType != Void.TYPE -> {
+          // List<Integer> pointToJson(Point point) {
+          val returnTypeAnnotations = method.jsonAnnotations
+          val qualifierAnnotations = parameterAnnotations[0].jsonAnnotations
+          val nullable = parameterAnnotations[0].hasNullable
+          object : AdapterMethod(
+            adaptersOffset = 1,
+            type = parameterTypes[0],
+            parameterCount = parameterTypes.size,
+            annotations = qualifierAnnotations,
+            adapter = adapter,
+            method = method,
+            nullable = nullable
+          ) {
+            private lateinit var delegate: JsonAdapter<Any>
+            override fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+              super.bind(moshi, factory)
+              delegate = if (Types.equals(parameterTypes[0], returnType) && qualifierAnnotations == returnTypeAnnotations) {
+                moshi.nextAdapter(factory, returnType, returnTypeAnnotations)
+              } else {
+                moshi.adapter(returnType, returnTypeAnnotations)
+              }
+            }
+
+            override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+              val intermediate = invoke(value)
+              delegate.toJson(writer, intermediate)
+            }
+          }
+        }
+        else -> {
+          throw IllegalArgumentException(
+            "Unexpected signature for $method.\n@ToJson method signatures may have one of the following structures:\n    <any access modifier> void toJson(JsonWriter writer, T value) throws <any>;\n    <any access modifier> void toJson(JsonWriter writer, T value, JsonAdapter<any> delegate, <any more delegates>) throws <any>;\n    <any access modifier> R toJson(T value) throws <any>;\n"
```

Comment:
Make this a raw string for readability

Reply:
I'm not sure how you wanted the indentation in the raw string?

Reply:
It just should have the same indent, your initial version had multiple inconsistent indents

Reply:
https://github.com/square/moshi/pull/1467/commits/a37c2bf5ea9480cf77255b4848d880195ea64707